### ISSUE TITLE
[FEATURE] Override/extend existing content elements/records

### ISSUE THREAD

**Issue by thommyhh** [FriendsOfTYPO3/content-blocks#34]
It would be nice to be able to extend existing records (pages, tt_content, ...) with new fields using the YAML syntax of the `content_blocks` extension.

I would suggest a folder structure like `ContentBlocks/Overrides/{ContentElements,PageTypes,RecordTypes}` similar to the `Configuration/TCA/Overrides` folder used by the TYPO3 Core.

If the table to be overridden does not exist - in case of `RecordTypes` - an exception should be thrown to make it clear, that there is an error.

The YAML syntax could look like this:

```yaml
name: t3core/1 # this should override doktype = 1; t3core as vendor prefix indicates, that this is a core type, which does not have a prefix in the TCA
fields:
   # Adds new field after the field `seo_title`
   -  identifier: my_special_field
      type: Textarea
      enableRichtext: true
      after: seo_title # Can be either a field or `palette:{name_of_palette}`
      linebreak: 'before' # Could be `before`, `after` or `both` to insert additional line break before and/or after the field we add
   # Adds new field before the field `description`
   -  identifier: my_other_field
      type: Link
      before: description # Can be either a field or `palette:{name_of_palette}`
      linebreak: 'after' # Could be `before`, `after` or `both` to insert additional line break before and/or after the field we add
   # Adds new tab at the end
   -  identifier: settings_tab
      type: Tab
      fields:
         -  identifier: header_menu_pages
            type: Relation
            allowed: pages
      # It is also possible to add a tab after field or palette like with a single field. The new tab is inserted after the tab, the given field or palette is located in
      after: no_follow # Field identifier or `palette:<name-of-palette>`
```

Another feature could be to make it possible to override existing fields, e.g. to assign a new label, change the configuration or move the field to another tab or palette. However, the type must not change, to avoid issues the underlying database field.
```yaml
name: t3core/1 # this should override doktype = 1; t3core as vendor prefix indicates, that this is a core type, which does not have a prefix in the TCA
fields:
   # This would throw an exception "You must not change the type of an existing field. Do not use `useExistingField` and `type` together."
   -  identifier: keywords
      useExistingField: true
      type: Textarea
      after: description 
      linebreak: 'before'
   # Move the field after the `description` field in the `SEO` tab
   -  identifier: abstract
      useExistingField: true
      after: description 
      linebreak: 'before'
```

**Comment by nhovratov** [comment]
Thanks for your feature request! This is indeed a common issue people have. I'm a little bit reluctant to mix Content Types defined natively with Content Block Overrides. My idea was to always operate on the same level. So Content Blocks can only override other Content Blocks. I have a feeling such a feature could stand for it's own and doesn't need to be part of "Core" Content Blocks. The problem is, right now there is no Content Blocks API which can read / analyse Core Types. Only Content Block Types are known to us. Of course it would be much easier if everything would be defined as a Content Block. But I don't see that coming soon.

What do you think? Maybe it's better suited as a separate project? Right now there are also other projects which can do such things (but not with YAML). For exmple: https://git.spooner.io/spooner-web/tcabuilder

**Comment by thommyhh** [comment]
You're welcome. I definitely agree with you, that it will take a long time to migrate from the current TCA definitions to Content Blocks for all record types, if it even happens.

Anyway, adding additional tabs, palettes and fields to existing core or third party record types is almost as common as defining custom content elements, which I see as the main purpose of content block. Therefore I think it would be a nice addition to this extension. The YAML syntax, the meaningful default values, the kind of scoped language files and the automatically generated SQL would make it a lot easier to add custom fields.

I do agree, that this overriding feature adds a whole new level. So we could put the whole reading of the TCA and the mapping to an object structure with tabs, palettes and fields into another project/extension. However, it would be great, if we still can use the TCA generation, scoping, SQL generation, et.al. from content blocks.

I came across `tcabuilder` before. It's nice to have kind of an API around the TCA. But content blocks really adds some convenience, which would make it much easier for integrators and developers alike.

**Comment by fe-hicking** [comment]
Just my 2cts, I hope I don't hijack this: A major pull for me to use content-blocks is, that what it does will be put into the core. Using other extensions on top of it or additionally sounds a bit "dangerous". As we all know, every deviation from the core is an uncertain way forward.

We discussed this topic internally today, too, and there was mixed feedback. We'd try to not extend core page/doktypes with content-blocks, but rather create custom page types and re-use palettes or fields of the core. Or if it must be, use the "old-school override".

However, as for Content Elements (not Page Types, but maybe sharing the conceptual problem?) what we'd like to be able is to create some sort of "core-sitepackage" that we can re-use for different projects, creating a "core slider". But then projects might add a little variation on top of that "core slider" to make it i.e. a "education slider" with distinct extra fields. In that case we'd like one content block (of a different extension, but the same vendor) to enhance these.

Scope is important, and creating a reasonable feature set. But maybe while a strategy is still up for discussion, ways to enable this can be found.

My colleague Marcel has some more in-depth ideas about this that we're trying to put together, but since this issue here opened up, I wanted to raise my hand, too 😅

**Comment by nhovratov** [comment]
Thank you both for your input!

Just to sum up the discussion so far:

- Extending Core or third-party Content Types is as common as creating new Content Types
- Right now, Content Blocks only operates on "Content Block"-level and doesn't know about other Types.
- Using third-party extensions for overrides (or using the native way) degrades the unified experience of Content Type management with Content Blocks.
- Creating custom Content Types should be the preferred way for customizing Types.
- We agree that overrides for other Content Blocks should be possible (core slider -> education slider + custom fields, custom template). -> this is already in my agenda and I would love to have some help here.

**Comment by sbuerk** [comment]
If I read the thread here, there are different points mentioned or wished. @nhovratov already stated most of them

About the point to `missuse` the content-block technique to manage TCA in general, I would disagree here.

Content-blocks should operate only for tt_content and content elements (with the exception for relation tables). 

Otherwise, if this should be done as a "general" move to define tca as yaml - the content-block would be wrong place for it.

I'm getting the point and whish for that, but I'm not convinced that that suits every people needs. There have been approaches for the core and TCA to improve it (not using yaml), at least at first to have a object(class based) structure after reading as a first step.

So that part needs to be thought in a bigger picture.

And opening content-blocks to use yaml to manipulte TCA in general is a cross-over which **needs** to be prevented.

**Comment by nhovratov** [comment]
Hey, @sbuerk. Content Blocks is not only for Content Elements, but also for Page Types and custom Record Types. And this is a good thing in my opinion. The quesion in the room is, whether it should expand into areas not defined within Content Blocks (=TCA manipulation). Right now it only generates **new** TCA. Only exception is the type field, where it adds its new types to the select.

**Comment by greenfieldr** [comment]
As exciting as I find the idea of changing native types, primarily in relation to PageTypes, I can fully understand that this is a barrier that we would rather not break through as long as it does not come to the point where all types are defined in this way in the core. 
Nevertheless, as Garvin has already described, we find the idea of bringing a basic set of content elements into our projects via kinda boilerplate extension and then being able to customize them in a project-specific sitepackage a very promising  concept that we would like to pursue further. @nhovratov, feel free to give Garvin and me a ping and we can see how we can help conceptually and/or with development time.

**Comment by nhovratov** [comment]
We encourage to create new types if the need arises to extend existing types. This already works for pages and tt_content. It is planned to support it for other (foreign) tables as well: https://github.com/nhovratov/content-blocks/issues/103
Otherwise, TCA overrides will always work.

**Comment by nhovratov** [comment]
It is now possible to create new types for native Record Types. This should be the preferred method.
